### DISCUSSION TITLE
Plans for fpdf2 to join the py-pdf org

### DISCUSSION THREAD

**Discussion by Lucas-C** [py-pdf/fpdf2#752] [Announcements]
Hi users of `fpdf2`! 😊

I am the current maintainer of `fpdf2`.
I started contributing PRs to this project during the summer of 2020.
At the time, @alexanderankin who created this fork of [PyFPDF](https://github.com/PyFPDF), was the maintainer.

Since then, he handed over maintainership to me, `fpdf2` evolved significantly, and I am quite proud of its current set of features, its stability and its performances (even if there is always room for improvements 😄).
And this was only possible thanks to the numerous, generous and skilled `fpdf2` contributors: https://github.com/PyFPDF/fpdf2#contributors-
Many thanks to you all! 👍🙏

However, I have been the only active maintainer for more than a year now, answering [issues](https://github.com/PyFPDF/fpdf2/issues) and making or reviewing [Pull Requests](https://github.com/PyFPDF/fpdf2/pulls).
While considering the future of `fpdf2`, and the possibility that I will have less time to dedicate to this project at some point,
I have been considering how to share the ownership of `fpdf2` with other benevolant developpers for several months now.

Last week, a discussion popped up in @sarnold fork of @pmaupin great [pdfrw](https://github.com/pmaupin/pdfrw) library, about how to maintain this fork and share ownership: https://github.com/sarnold/pdfrw/issues/15. One idea was to move this fork to the @py-pdf GitHub org, managed by @MasterOdin & @MartinThoma, who maintain https://github.com/py-pdf/pypdf, the successor of `PyPDF2`.
At the same time, @reingart reached out to me & @MartinThoma by email about archiving or handing over https://github.com/reingart/pyfpdf.
Given this context, and the presence of some motivated Python developpers willing to share efforts to maintain libraries dedicated to parsing & building PDF documents, I think this is the right opportunity to join forces!

Hence this discussion, that aims to present my current plans for the future of `fpdf2` and get feedback from you, the `fpdf2` community 😊

## ✅ Step 1 - Establish some basic guidelines for maintainers of the py-pdf org
The goal is to establish some basic principles for maintainers to work together & help each other.

For now I have mostly in mind some basic rules on how to manage **releases** ("when ?" & "how ?", for example [fpdf2 has a checklist](https://pyfpdf.github.io/fpdf2/Development.html#release-checklist)), and how people can can **become maintainers** For example, does it require explicit approval of all existing maintainers?

Basically, I initially just want to limit the risk of "broken" `fpdf2` releases being made an impacting users.
Those guidelines could then evolve over time.

I'm planning to submit a first draft as an issue of https://github.com/py-pdf/py-pdf.github.io for @MasterOdin & @MartinThoma to review it, and to take inspiration from the following existing guidelines / group of maintainers:
* OpenSource.guide: [Best practices for Maintainers](https://opensource.guide/best-practices/) & [Leadership and Governance](https://opensource.guide/leadership-and-governance/)
* https://jazzband.co

## ✅ Step 2 - Move fpdf2 to https://github.com/py-pdf
This should be relatively straightforward, as GitHub makes this relatively easy:
https://docs.github.com/en/repositories/creating-and-managing-repositories/transferring-a-repository

At this point, all @py-pdf maintainers will be able to handle `fpdf2` issues & PRs, and perform new releases of the library.

## Step 3 - Homogenization
Personnally, I'd like to contribute to https://github.com/py-pdf/pypdf (and https://github.com/sarnold/pdfrw if it joins @py-pdf) in order to discover & learn about how they manage things: code structure, linters, unit tests...

And in the end, I have the idea of sharing some practices toward static code analysis, tests & CI/CD pipelines between projects of the @py-pdf organization. To make things easier for @py-pdf maintainers and also because I like setting up this kind of stuff 😅

## Community feedbacks
Do you have any questions regarding this plan?
I'd love to get your reactions! 😊

_Ping to some major & recent contributors: @alexp1917, @gmischler, @eroux, @erap129, @Bubbu0129, @RedShy, @andersonhc, @RubendeBruin, @ssavi-ict_

**Comment by MartinThoma** [comment]
> At this point, all https://github.com/py-pdf maintainers will be able to handle fpdf2 issues & PRs, and perform new releases of the library.

Correct, due to https://github.com/PyFPDF/fpdf2/blob/master/.github/workflows/continuous-integration-workflow.yml#L89-L101

Please note that this does not give anybody else access to PyPI. The current owners within PyPI still stay the (only) owners. Meaning if one of the GitHub py-pdf admins went evil / leaked their credentials, it would be possible to clean up the mess.

Let's hope that never happens though :crossed_fingers:

**Reply by MartinThoma** [reply]
@Lucas-C By the way, that might actually be something that pypdf would take over from fpdf2 :-) I was hesitant to give anything / anybody else access to PyPI for security concerns. However, I now understand better how it works and I think I would notice malicious uploads + GitHub is mostly trustworthy ([errors happen, though](https://securityboulevard.com/2023/04/after-brief-exposure-in-public-repo-github-rotated-private-ssh-key/))

**Comment by ssavi-ict** [comment]
Thank You @Lucas-C for tagging.
> Step 2 - Move fpdf2 to https://github.com/py-pdf

Out of curiosity, what will be the advantages of moving `fpdf2` to `https://github.com/py-pdf`? Apologies for not getting things as fast as possible.

**Reply by Lucas-C** [reply]
> Out of curiosity, what will be the advantages of moving `fpdf2` to `https://github.com/py-pdf`? Apologies for not getting things as fast as possible.

No worries, thanks for your feedback and question 😊

The main drive is to share maintainer efforts: current members or @py-pdf will be able to answer `fpdf2`-related issues and even perform new releases, and I will be able to help on the other projects of the @py-pdf org.

**Reply by MartinThoma** [reply]
I see a couple of advantages (going in a similar direction):

1. **Signaling**: People who are not that deep into the Python/PDF ecosystem have an easier time to judge if they can rely on that project receiving updates. Seeing that this is not the toy project of a single developer, but that the core developer(s) decided to put some effort in project governance is a huge plus from a users perspective (for some professional users at least xD). It becomes less likely that the package is deleted out of protest (see the Azer Koçulu story).
2. "**Advertisement**": Some people know about pypdf, but not about fpdf or vice versa. Getting a stronger collaboration might help to push the other projectes.
3. **Development**: I hope that we can exchange design ideas / look into the partner projects and see how they solved some PDF-specific issues. For example, PDF forms is a topic where pypdf would profit a lot from other developers insights. Recently the new table-feature was added to fpdf2 - maybe in future pypdf developers can add an amendment so that fpdf2 automatically generates semantic tags.
4. "**Live insurance**": pypdf / PyPDF2 / PyPDF3 / PyPDF4 were dead for a long time. It was a good amount of luck that the original maintainers of pypdf and PyPDF2 were still reachable + willing to transfer their projects. If we have more people who have proven to be trustworthy, the project becomes less dependent on single developers. This is one aspect of the "signaling" argument.
5. **Funding**: I'm currently looking into solutions to allow people to donate for pypdf. The first goal is to buy the PDF 2.0 standard. At very least we can share the experience how that works / how much work it is to set up. Maybe we can even say that the funding is for the py-pdf organization, not for the pypdf project.
6. **Combining codebases**: That is nothing I'm actively planning, but I would be happy if we would see over time a common core. Maybe just parts for testing. Or maybe pypdf could use fpdf2 to generate some of the test cases.

To summarize: It's definitely not necessary to merge. But both projects might benefit from being in the same Github organization

**Reply by ssavi-ict** [reply]
@MartinThoma Thank You for the detailed explanation.

@Lucas-C Thank You too. 
I would like to mention one thing. We can use some comments in each file especially for `tutorial`, `doc` etc. type directories so that we can track the dependencies easily, which I think will be helpful for the beginner contributors like me. I had a conversation with you about restructuring the `docs` folder of `fpdf2`. I am now feeling something similar is needed.

**Reply by gmischler** [reply]
> 6\. **Combining codebases**: That is nothing I'm actively planning, but I would be happy if we would see over time a common core. Maybe just parts for testing. Or maybe pypdf could use fpdf2 to generate some of the test cases.

This touches one of the first questions that popped into my mind when I saw this proposal.

Merging the code of those projects is obviously not realistic in the short term.
But if they get joined under one organisational umbrella, then there should be a clear idea about which role each one is expected to play.
While fpdf2, PyPDF, and pdfrw have been created with different goals in mind, there is still a noticeable overlap in functionality.

So how can we best differentiate them? Actually, this question seems to also include [py-pdf/PyPDF-Builder](https://github.com/py-pdf/PyPDF-Builder) and [py-pdf/pdf](https://github.com/py-pdf/pdf).

As a first step, maybe a short list of the available tools and their roles can be included in their individual READMEs.
When searching online for Python PDF libraries, the returned selection is rather confusing and includes heaps of long obsolete options. Combining a number of active projects in one organisation should help to reduce that confusion, not increase it... :wink:

In the long term, it might indeed be interesting to consider whether some functionality common to several of our candidates can be reasonably farmed out to a common core library. But I don't think we need to worry about that very much for now.

**Reply by Lucas-C** [reply]
> As a first step, maybe a short list of the available tools and their roles can be included in their individual READMEs.
> When searching online for Python PDF libraries, the returned selection is rather confusing and includes heaps of long obsolete options. Combining a number of active projects in one organisation should help to reduce that confusion, not increase it... 😉

Thanks for the feedback @gmischler!

I really don't think that grouping those repositories in a single GitHub org will increase confusion 😊
It can only makes things more coherent in the long term.

Regarding having a list of available tools, we already have several sections with such lists:
* https://github.com/py-pdf/awesome-pdf
* https://pyfpdf.github.io/fpdf2/#related
* https://github.com/pmaupin/pdfrw#other-libraries

In the end, we should probably regroup those 3 lists on https://github.com/py-pdf/py-pdf.github.io,
with a clear decription of the ìntended usages of `pypdf` / `pdfrw` / `fpdf2`

**Reply by gmischler** [reply]
> I really don't think that grouping those repositories in a single GitHub org will increase confusion

It actually may.

If I'm looking at packages from different sources, then the probability of me looking deeper in the documentation for a "related" list is definitively there. I remember finding fpdf2 that way when looking for a tool less than two years ago, and it was a distinctly non-trivial search.

But when several tools are homed under a common roof, then I expect to find a list of their  different purposes/target audiences/other distinctions right at the top. And since several of the packages have a rather complex genealogy, info should also be included as to which other (possibly outdated/obsolete) packages they are meant to replace.

This kind of information is simply good "customer service", and to some degree "marketing". You want to make it as easy as possible for prospective users to find the right tool for their task.

**Reply by Lucas-C** [reply]
> But when several tools are homed under a common roof, then I expect to find a list of their different purposes/target audiences/other distinctions right at the top. And since several of the packages have a rather complex genealogy, info should also be included as to which other (possibly outdated/obsolete) packages they are meant to replace.

I agree overall 😊
It's just not my top priority a the moment.

Would you want to contribute some documentation toward this valuable goal @gmischler?
I think you can already start a draft for such a page on https://github.com/py-pdf/py-pdf.github.io, with some initial description of those 3 libraries and their history

**Reply by gmischler** [reply]
> It's just not my top priority a the moment.

At some point (after the move), having the goals and target audiences of the various projects clearly stated may also help in deciding the future development direction of each of them, or a potential common core library.

Maybe in the future, some thought could also be invested in finding more memorable and distinguishable project names. Or am I the only one who keeps getting confused by names in the form of `"(py|f)?pdf[2345]?"` modulo all possible case variations? 😕 (As a related aside: Usurping a generic term like in [py-pdf/pdf](https://github.com/py-pdf/pdf) is something the Microsoft marketing department could have come up with...)

> Would you want to contribute some documentation toward this valuable goal @gmischler?

I'm not particularly familiar with most of the projects involved, let alone their history. I just couldn't help but notice that there is a lot of "heritage" all around. Descriptions of the distinguishing characteristics should probably rather be crafted by people intimately familiar with each individual project.

**Reply by MartinThoma** [reply]
Here are some pages that compare the different projects:

* [pypdf vs X](https://pypdf.readthedocs.io/en/latest/meta/comparisons.html) written by me
* [The Python PDF Ecosystem in 2023](https://martinthoma.medium.com/the-python-pdf-ecosystem-in-2023-819141977442) written by me
* [pikepdf README](https://github.com/pikepdf/pikepdf)
* [The fpdf FAQ](https://pyfpdf.readthedocs.io/en/latest/FAQ/index.html)

We should probably make one page about the Python / PDF ecosystem on https://py-pdf.github.io/. Then our projects can link to that page.

**Reply by MartinThoma** [reply]
> getting confused by names in the form of `"(py|f)?pdf[2345]?"`

That is a complicated discussion for existing projects. The decision here should only be at the project maintainers, not in the py-pdf organization.

Besides that: I completely agree! At least for pypdf I hope it's simple now: All lowercase, no numbers. But moving the project is a major step. I wouldn't want to do it again for pypdf.

I would recommend to keep the following consistent and always make all-lowercase names:

1. **Package name on PyPI**
2. **Imported package** (e.g. you install `PyMuPDF` but import `fitz`). Having a different name only makes sense to me for drop-in replacements (PIL and pillow). So it could make sense to have fpdf2 as the name on PyPI and fpdf as the imported name - but such a decision should only be made at the beginning of a project, not after one has a significant user base.
3. **Communicated name**: Here I sometimes see [PyFPDF](https://pyfpdf.readthedocs.io/en/latest/Tutorial/index.html) which I perceive as confusing. I guess the reason was to distinguish between the original FPDF (PHP?) and the Python version. [fpdf2 does it well](https://pyfpdf.github.io/fpdf2/)

**Comment by andersonhc** [comment]
As long as the fpdf2 project can keep it's characteristics of being welcoming and encouraging new contributors, I don't see a problem.

**Reply by ssavi-ict** [reply]
@andersonhc Correct. I also feel the same. Contributing to this project feels easy from some other projects that I explored.

**Reply by Lucas-C** [reply]
Thank you for your kind words @andersonhc & @ssavi-ict 😊
Makes my day!

**Reply by marco-doerig** [reply]
I can only agree with the previous speakers. After a long search (because it is really opaque with these pdf libraries) I stumbled upon this repo and am happy to have found it. You take good care of it and I really hope that after a possible transfer it will continue.

**Comment by RubendeBruin** [comment]
@Lucas-C , first of all thanks for all the work you have been putting in. Maintaining a package like fpdf2 on your own is fun and rewarding but can easily become a burden and a responsibility. I think it is good that you found a way to share this with multiple people. Although shared responsibility is no responsibility and when creating something is it often easier to be able to dictate rather than having to get approval from others. 

For users like me the combination of fpdf2 and pypdf is great as I'm using both of them together: fpdf2 to export some chapters and pypdf to mix in drawings (from autocad and such) and appendices. Having both the tools supported by the same people would be a good thing.

Next step could be merging/refactoring the packages such that they use common resources from the same pool. Logical candidates would be test-files, CI, testing and documentation.
But there is also a good possibility that the two packages are too fundamentally different to share common parts. Especially with python being the free language that it is it is quite easy to develop the same things in completely different ways and bringing them in-line can me more trouble than it is worth. That is not an issue as long as the use-cases are not overlapping (which I think is the case) although working with different styles can be annoying if you are working on both packages. 

TLDR: YES :-)

**Comment by MartinThoma** [comment]
Ownership of the py-pdf organization is another important aspect. I want to make @Lucas-C an owner, but I also want the community to participate in the discussion. I've linked it already for the pypdf / PyPDF2 people, but maybe you're interested as well: https://github.com/py-pdf/py-pdf.github.io/discussions/3 :-)

**Comment by gmischler** [comment]
Btw: Kind of funny in this context that fpdf2 is not even mentioned (yet) on [awsome-pdf](https://github.com/py-pdf/awesome-pdf)... 🤨

**Reply by MartinThoma** [reply]
I've added it :smile: 

By the way: Anybody can add PRs there ... or open an issue. It was coincidence that I've seen your comment here.

**Comment by MartinThoma** [comment]
As a side-note: We now use fpdf2 in `pypdf` (the project, not the organization) in CI: https://github.com/py-pdf/pypdf/pull/1907/files

I want to extend this kind of usage in future :-)

**Reply by Lucas-C** [reply]
Nice!

Regarding this topic of moving `fpdf2` to the @py-pdf org,
on my side this process is currently pending on PR https://github.com/py-pdf/py-pdf.github.io/pull/1

**Reply by MartinThoma** [reply]
thanks for letting me know - I didn't see that PR :see_no_evil:  I moved in the middle of May and missed a couple of things during that time :sweat_smile:

**Comment by Lucas-C** [comment]
I have just moved `PyFPDF/fpdf2` to https://github.com/py-pdf/fpdf2,
after releasing a new version: https://github.com/py-pdf/fpdf2/releases/tag/2.7.5

**Reply by Lucas-C** [reply]
The PyFPDF GitHub organization is now **archived**: https://github.com/PyFPDF

And I have setup a redirection from https://pyfpdf.github.io/fpdf2/ to https://py-pdf.github.io/fpdf2/,
_cf._ https://github.com/PyFPDF/pyfpdf.github.io

**Comment by ramanasai** [comment]
I need Help in Header where i want the header text to be in all pages and the header text changes dynamically in some pages is this possible

**Reply by gmischler** [reply]
@ramanasai, please open a new thread for your question. You are completely off-topic in this specific discussion here.

**Reply by Lucas-C** [reply]
Yes, I second @gmischler answer.
Please open a dedicated [issue (click there)](https://github.com/py-pdf/fpdf2/issues/new/choose)
or [discussion (click there)](https://github.com/py-pdf/fpdf2/discussions/new/choose), and we will try to help you.

Please also take care to provide some code demonstrating what your are trying to achieve,
otherwise we won't be able to help you much 😊
